cranelift/wasm/src/code_translator/bounds_checks.rs

```diff
@@ -369,26 +422,102 @@ fn explicit_check_oob_condition_and_compute_addr(
 /// It is the caller's responsibility to ensure that any necessary bounds and
 /// overflow checks are emitted, and that the resulting address is never used
 /// unless they succeed.
-fn compute_addr(
-    pos: &mut FuncCursor,
+fn compute_addr<Env>(
+    // pos: &mut FuncCursor,
     heap: &HeapData,
     addr_ty: ir::Type,
     index: ir::Value,
     offset: u32,
-) -> ir::Value {
+    builder: &mut FunctionBuilder,
+    env: &mut Env,
+) -> WasmResult<ir::Value>
+where
+    Env: FuncEnvironment + ?Sized,
+{
+    use cranelift_codegen::constants::{
+        BIT_MASK_EXCLUDING_MTE_TAG, BIT_MASK_INCLUDING_MTE_TAG, MTE_DEFAULT_FREE_TAG,
+        MTE_TAG_LEAST_SIG_BIT,
+    };
+
+    let pos = builder.cursor();
     debug_assert_eq!(pos.func.dfg.value_type(index), addr_ty);
 
-    let heap_base = pos.ins().global_value(addr_ty, heap.base);
-    let base_and_index = pos.ins().iadd(heap_base, index);
-    if offset == 0 {
+    // We expect that the heap base address already contains the
+    // MTE_LINEAR_MEMORY_FREE_TAG tag.
+
+    // TODO: we replaced the line below with the line below that. Was that correct?
+    // let heap_base = pos.ins().global_value(addr_ty, heap.base);
+    let heap_base = builder.ins().global_value(addr_ty, heap.base);
+
+    // TODO: I think there might be more efficient hardware instructions for this (checkout CMPP and SUBPS)
+
+    // If the index has already been tagged (by e.g. segment_* instructions),
+    // then we have to remove the tag stored in the heap base address.
+    // Otherwise, both the base address and the index would have tags, leading
+    // to arithmetic errors during the addition.
+    //
+    // Pseudo code:
+    //
+    // if (index_tag != 0) {
+    //     [INDEX_IS_TAGGED_BLOCK]
+    //     // Remove tag from heap_base
+    //     heap_base = heap_base & tag_bits_exclusive;
+    // }
+    // [NEXT_BLOCK]
```

Comment:
This definitely has a lot of room for optimization. As explained above, we have to check whether the index is already tagged, because our `MTE_LINEAR_MEMORY_FREE_TAG` tag is currently `0b0001`. If we don't perform this check, and we would add the index onto the base_ptr normally, the tag of the sum would be `1 + <index tag>`, even though it should be `<index tag>`. But this if condition could become quite expensive.
So one easy optimization I thought of: Why don't we just choose `0b0000` as the `MTE_LINEAR_MEMORY_FREE_TAG` (and then `0b0001` as the `MTE_DEFAULT_FREE_TAG`, which we would then have to set in Linux as well to ensure Linux assigns that tag to addresses by default)? This would result in `0 + <index tag>`, which is just `<index tag>` as the tag on the sum. Anything I am missing here?